### PR TITLE
feat: add `FormattedRelativeTime` component & `formatRelativeTime`

### LINKED ISSUE #443: Server/Client time correctness
Many Windows users do not sync their OS clocks as do OSX users autom. Very often Relative times will render time differences that are too big for a good User Experience...

A quick example of a random UI rendering/data flow:

`Date.now()` -> generates -> `Message sent 1 second ago` -> Server responds > `{created_at: <Date>}` -> React UI refreshes value -> `Message sent 35 seconds ago` (respond took 2 secs).

Not long ago `initialNow` was introduced, would it make sense to add something like `serverNow`?

### PR DESCRIPTION
**Pull request by longlho**
should also fix #443

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -293,6 +293,31 @@ By default, the `value` is compared with the current time at the time the functi
 
 **Note:** The reason [`IntlShape`](#Intlshape) has a `now` function is to allow both `<IntlProvider>` and `<FormattedRelative>` components to provide an `initialNow` prop. This allows for the current time to be fixed for things like testing or server-side rendering in an isomorphic/universal React app.
 
+#### `formatRelativeTime`
```
Comment (redonkulus) [inline docs/API.md]: We should differentiate how FormatRelativeTime is different from FormatRelative.

File: docs/API.md [hunk 2]
```diff
@@ -293,6 +293,31 @@ By default, the `value` is compared with the current time at the time the functi
...
+    now?: number;
+  }
+): string;
+```
+
+This function will return a formatted relative time string (e.g., "1 hour ago"). It expects a `value` which is a number, a `unit` (`second`, `minute`...) and accepts `options` that conform to `Intl.RelativeTimeFormatOptions`.
+
+```ts
+formatRelativeTime(0); // "now"
+formatRelative(-1, 'second'); // "1 second ago"
```
Comment (redonkulus) [inline docs/API.md]: shouldn’t this be formatRelativeTime for all examples.

File: docs/Components.md [hunk 3]
```diff
@@ -263,6 +264,47 @@ By default `<FormattedRelative>` will render the formatted relative time into a
 
 **Note:** You can adjust the maximum interval that the component will re-render by setting the `updateInterval` in `ms`. A falsy value will turn off auto-updating. The updating is smart and will schedule the next update for the next _interesting moment_.
 
+### `<FormattedRelativeTime>`
+
+This component uses the [`formatRelativeTime`](API.md#formatrelativetime) API and has `props` that correspond to the following relative formatting options:
```
Comment (redonkulus) [inline docs/Components.md]: Again we should differentiate this

**Comment by DragonRaider5** [comment]
Excuse my lack of knowledge, but what exactly is the difference?
Don't do two components for the same use case confuse users?

**Comment by longlho** [comment]
Overall `formatRelative` & `FormattedRelative` have strayed from the [spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RelativeTimeFormat) which makes their behavior more high-level than they used to be:
- `formatRelative` as a function uses `formatRelativeTime` under the hood, but does the `value - now` calculation automatically & `best-fit`. While this is a feature, it suffers from things like `Date` parsing issue being very inconsistent across browsers, clock skew & timezone issues. What `best fit` does is whenever the time difference crosses a threshold, it changes the `unit` internally. Over at [formatjs monorepo](https://github.com/formatjs/formatjs) we already got a bunch of issues like "does 13 hour difference count as a different day?" which is subject to timezone.
- `FormattedRelative` suffers from the same set of issues. It has the auto-update feature which we should keep.

Overall they no longer follow the spec, which makes their options confusing since `style` is something different now. `FormattedRelativeTime` & `formatRelativeTime` are spec-compliant. We should keep the extra features of the old ones but either rename them, or fold them into `FormattedRelativeTime` instead.

**Comment by longlho** [comment]
Another interesting tidbit/compatibility issue is that `-0` is rendered as `0 seconds ago` & `+0` is `in 0 seconds`, whereas in `formatRelative` right now there's no way to distinguish

**Comment by longlho** [comment]
So I've been giving it some thoughts and we can merge the 2 with the following changes:
- Rename `FormattedRelative` to `FormattedRelativeTime` bc the native API is called `Intl.RelativeTimeFormat`
- Same for `formatRelative` -> `formatRelativeTime`
- Change `value` prop to difference (0, -1, 1...) instead of absolute time to be spec-compliant
- Introduce a new prop called `date` that takes in absolute `Date`
- Change `style` to be spec-compliant (`long/short/narrow`)
- Add `numeric` prop to be spec-compliant (`always/auto`)
- Add `bestFit` as a prop that determines if the component should "auto-promotes" unit (e.g 60 seconds -> 1 minute).

Overall it's something like:
```tsx
<FormattedRelativeTime date={new Date()} /> // now
// same as
<FormattedRelativeTime value={0} unit="second" /> // now

<FormattedRelativeTime date={new Date()} unit="hour" /> // this hour
// same as
<FormattedRelativeTime value={0} unit="hour" /> // this hour

<FormattedRelativeTime date={new Date()} bestFit /> // now

<FormattedRelativeTime date={Date.now() + 60000} bestFit /> // in 1 minute
// same as
<FormattedRelativeTime value={1} unit="minute" /> // in 1 minute

<FormattedRelativeTime date={Date.now() + 60000} /> // in 60 seconds
// same as
<FormattedRelativeTime value={60} unit="second" /> // in 60 seconds
```

Thoughts?

**Comment by DragonRaider5** [comment]
Sweet! :+1:

**Comment by DragonRaider5** [comment]
Is there something I could help with?

**Comment by longlho** [comment]
@DragonRaider5 	thanks! I pinged u over at #1328 1328

**Comment by DragonRaider5** [comment]
Yeah, I can take care of this tomorrow or the like :+1:

However my question originally was aiming at the current changes, as I know the updating mechanisms fairly well.
We'll obviously need to disable them in case `value` is passed. We should also throw an `Invariant` in case both `value` and `date` are being passed.

Also need to update the Docs.

**Comment by longlho** [comment]
Yup I'll have to make multiple PRs for this otherwise it'll get too big.

**Comment by DragonRaider5** [comment]
Do you think so?

Todo (from what I understood):
- [ ] Invariant in case both `value` and `date` are set
- [ ] Only enable auto-update when `date` is set
- [ ] Rename component/function
- [ ] Change types
- [ ] Update docs

Aren't multiple PRs for one feature kind of messy?

**Comment by longlho** [comment]
hmm you're prob right, maybe 1 PR can fit.

**Comment by longlho** [comment]
closes in favor of #1338

**Pull request closed without merging**